libnmstate/state.py

```diff
@@ -167,6 +226,49 @@ def merge_interfaces(self, other_state):
             dict_update(other_state.interfaces[name], self.interfaces[name])
             self._ifaces_state[name] = other_state.interfaces[name]
 
+    def merge_route_config(self, current):
+        """
+        Merge routes from current to self.
+        If any changed route referring non-exist interface, create the
+        interface state with name only, in order to trigger the recreating of
+        the interface profile.
+        """
+        # Merge other_routes
+        iface_route_sets = defaultdict(set)
```

Comment:
I am adding this comment after reading the implementation below.

Looks like you actually need indexed routes, where the entries themselves are the `_Route` object.
So, you could use that in the indexed version you currently use (convert the route entries into the object on state creation in the first place).
The `set` is also something that can be maintain from the beginning: Use a set instead of a list in the indexed structure.
Then, you just need to read it from `self` and `other`.
Note that even if you do not do that (for whatever reason), having these structures created for you by a state method fits here: `self._indexed_route_objects_by_iface().

If we separate between the self and other data/sets, then you can do whatever calculation you want in the end:
```
common = self._indexed_routes & other._indexed_routes 
all = self._indexed_routes | other._indexed_routes
...
```

Reply:
The `desired_state.iface_routes` is used by `metadata.py` which does not need to know the route
object set stuff.

Doing the converting to object sets in `__init__()` or `merge_route_config()` has no difference right now as `merge_route_config()` is the only user for those route object sets.

Changing internal information from dict to object based is another big topic which I intend to do it in another PR.

Reply:
The internal representation of the data is an nmstate thing and you created an object to represent a route entry.
I do not understand your argument about why it is not used all over if it is useful.
If you have two data structures, it makes things more complex.

The claim that only the metadata is using something is not a reason to duplicate data or have two structures. If one can serve both, it makes things simpler and cleaner.

This is not a blocker, but I would prefer to see it done sooner than latter.